### PR TITLE
Feature/984: bulk update notification email addresses

### LINKED ISSUE #984: [FEATURE] bulk update notification email addresses
related to #983 

idea is, we give it a BU, a type and optionally keys
we then need to supply the email address(es) we want the message to be send to, an optional message (not relevant for all types) and the script then goes and sets it for all relevant keys and types

for those types that have 2 options of notifications (automation with Complete and Error) we need to be able to supply different values for the email address(es) and messages.

the log should clearly show what the previous value was or if no change was necessary (and then skip the update callout)

### LINKED ISSUE #1234: [TASK] investigate minimum scope / permissions required to run mcdev
**REST**: https://developer.salesforce.com/docs/marketing/marketing-cloud/guide/rest-permissions-and-scopes.html
**todo:** find similar overview for SOAP

### PR DESCRIPTION
**Pull request by yuliialikhyt**
# PR details

## What changes did you make? (Give an overview)

- closes #984

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] test scripts updated
- [x] Wiki updated (if applicable)

### CODE REVIEW

File: lib/cli.js [hunk 1]
```diff
@@ -543,6 +543,53 @@ yargs
             Mcdev.fixKeys(argv.BU, argv.TYPE, csvToArray(argv.KEY));
         },
     })
+    .command({
+        command: 'updateNotifications <BU> <TYPE> [KEY]',
+        aliases: ['updNotif'],
```
Comment (JoernBerkefeld) [inline lib/cli.js:548]: how about "un" as alias?
```suggestion
        aliases: ['un'],
```

File: lib/cli.js [hunk 2]
```diff
@@ -543,6 +543,53 @@ yargs
...
+                })
+                .positional('TYPE', {
+                    type: 'string',
+                    describe: 'metadata type',
+                })
+                .positional('KEY', {
+                    type: 'string',
+                    describe: 'key(s) of the metadata component(s)',
+                })
+                .positional('emailError', {
```
Comment (JoernBerkefeld) [inline lib/cli.js:564]: lets turn this into an option - better fit for the different requirements of the types because otherwise you couldn't "skip" a parameter: imagine somebody only wants completion notifications for automation with a custom note.

File: lib/cli.js [hunk 3]
```diff
@@ -543,6 +543,53 @@ yargs
...
+                    type: 'string',
+                    describe: 'email to notify that an error occured during execution',
+                })
+                .positional('emailComplete', {
```
Comment (JoernBerkefeld) [inline lib/cli.js:568]: lets turn this into an option - better fit for the different requirements of the types because otherwise you couldn't "skip" a parameter: imagine somebody only wants completion notifications for automation with a custom note.

File: lib/cli.js [hunk 4]
```diff
@@ -543,6 +543,53 @@ yargs
...
+                    type: 'string',
+                    describe: 'email to notify about successfull completion',
+                })
+                .positional('errorNote', {
```
Comment (JoernBerkefeld) [inline lib/cli.js:572]: lets turn this into an option - better fit for the different requirements of the types because otherwise you couldn't "skip" a parameter

File: lib/cli.js [hunk 5]
```diff
@@ -543,6 +543,53 @@ yargs
...
+                    type: 'string',
+                    describe: 'run error note',
+                })
+                .positional('completionNote', {
```
Comment (JoernBerkefeld) [inline lib/cli.js:576]: lets turn this into an option - better fit for the different requirements of the types because otherwise you couldn't "skip" a parameter: imagine somebody only wants completion notifications for automation with a custom note.

File: lib/cli.js [hunk 6]
```diff
@@ -543,6 +543,53 @@ yargs
...
+                    type: 'string',
+                    describe: 'run completion note',
+                })
+                .option('like', {
+                    type: 'string',
+                    group: 'Options for update notification email address:',
```
Comment (JoernBerkefeld) [inline lib/cli.js:582]:
```suggestion
                    group: 'Options for updateNotifications:',
```

**Comment by phjulia** [comment]
TODO:
1) Test method
2) function for import file
3) check if there are any other types apart for automation and import file where notifications can be set

File: lib/index.js [hunk 7]
```diff
@@ -1126,6 +1130,85 @@ class Mcdev {
...
     }
+    /**
+     * Updates notification email address field
+     *
+     * @param {string} businessUnit name of BU
+     * @param {TYPE.SupportedMetadataTypes} type limit execution to given metadata type
+     * @param {string[]} [keys] customerkey of the metadata
+     * @returns {Promise.<Object.<string, string[]>>} key: business unit name, value: list of affected item keys
+     */
+    static async updateNotifications(businessUnit, type, keys) {
```
Comment (JoernBerkefeld) [inline lib/index.js:1141]: - [ ] lets please include this in `Mcdev.#runMethod` like I did for `Mcdev.fixKeys`.

File: lib/index.js [hunk 8]
```diff
@@ -1126,6 +1130,85 @@ class Mcdev {
...
+    }
+    /**
+     *
+     * @param {string} cred name of Credential
+     * @param {string} bu name of BU
+     * @param {string} type metadata type
+     * @param {string[]} [keys] limit retrieval to given metadata keys
+     * @returns {Promise.<string[]>} keys of the automations where notifications were updated
+     */
+    static async #updateNotificationsBU(cred, bu, type, keys) {
```
Comment (JoernBerkefeld) [inline lib/index.js:1183]: - [ ] lets please include this in `Mcdev.#runOnBU` like I did for `Mcdev.#fixKeys`.

File: lib/metadataTypes/Automation.js [hunk 9]
```diff
@@ -1558,6 +1558,116 @@ class Automation extends MetadataType {
...
+        this.definition.bodyIteratorField = iteratorBackup;
+        return automationLegacyKeys;
+    }
+    /**
+     * A function to update automation email notifications
+     *
+     * @param {string} keys metadata keys
+     * @returns {Promise.<string[]>} keys of the automations where notifications were updated
+     */
+    static async updateNotifications(keys) {
```
Comment (JoernBerkefeld) [inline lib/metadataTypes/Automation.js:1585]: any publically accessible method on the type classes needs to be added as a placeholder on `metadataTypes/MetadataType.js` as well. Please check how that was done for `MetadataType.execute` and others.

File: lib/metadataTypes/Automation.js [hunk 10]
```diff
@@ -1558,6 +1558,116 @@ class Automation extends MetadataType {
...
+                                    notification.notificationType === 'Error' &&
+                                    (emailError !== notification.definition ||
+                                        Util.OPTIONS.errorNote !== notification.body)
+                            )
+                    )
+                ) {
+                    // create payload
+                    const notificationBody = {
+                        programId: automationLegacyMapObj.metadata[key].id,
+                        workers: [
```
Comment (JoernBerkefeld) [inline lib/metadataTypes/Automation.js:1625]: this does not consider that 
- the user could choose to only update one of the 4 fields (2 email fields, 2 note fields)
- the user could try to update the note for error OR completion but provide no email --- AND that no corresponding email was set before
-  the user might want to REMOVE completion or error notification completely
- ... likely more variants in the same way. basically, we should provide full flexibility here.

File: lib/metadataTypes/Automation.js (hunk 9 above)
Comment (JoernBerkefeld) [reply lib/metadataTypes/Automation.js:1585]: looking further, i noticed you did add a method called `setNotifEmailAddr`.
I'm guessing that was an earlier name for `updateNotifications` and only needs to be updated? :)

**Review by JoernBerkefeld — CHANGES_REQUESTED** [review]
i know this is still in draft but I wanted to give you a few pointers early on :)

**Comment by phjulia** [comment]
TODO:

- [ ] email validation in interactive questions and in updateNotifications itself
- [ ] a test for importFile updateNotifications, additional tests for different cases (same email address provided etc)

**Comment by JoernBerkefeld** [comment]
DataVerification activities also have a notification option... but u need to know their id to update them

**Comment by phjulia** [comment]
> DataVerification activities also have a notification option... but u need to know their id to update them

mcdev does not have data verification as a type yet. I would need to retrieve, then update and deploy like with importFile activity
Maybe in another PR?

**Comment by phjulia** [comment]
- [ ] function in Automation.js is too bulky, I think some parts can be moved into a helper method

File: lib/metadataTypes/Automation.js (hunk 10 above)
Comment (phjulia) [reply lib/metadataTypes/Automation.js]: Now it's possible to only update either of the notes (if there are email addresses), either completion email or error email or both. --clear option allows to remove all notes and emails, only notes, only either type of emails. 

If someone tries to set a note and there is no email address, the user is asked if they want to set email address

**Comment by phjulia** [comment]
@JoernBerkefeld I think it's ready for review. I will prepare documentation later today :)

**Comment by JoernBerkefeld** [comment]
on DataVerification:
yes, that's only in a new PR that I'm working on. we might want to create a new issue to cover notification updates for that type...

**Review by JoernBerkefeld — CHANGES_REQUESTED** [review]
we can have a call if you want?

File: test/type.automation.test.js [hunk 11]
```diff
@@ -883,4 +887,121 @@ describe('type: automation', () => {
             return;
         });
     });
+    describe('Update notifications ================', () => {
```
[the remark this answers is not in the document]
Comment (phjulia) [reply test/type.automation.test.js:890]: Added more tests for all of those cases + one negative test (invalid email)

File: lib/util/cli.js [hunk 12]
```diff
@@ -53,7 +53,33 @@ const Cli = {
...
+    /**
+     * interactive helper to set automation run completion/error note
+     *
+     * @param {TYPE.SupportedMetadataTypes} type note type (error/completion)
+     * @returns {Promise.<string[]>} responses
+     */
+    async updateNotificationEmails(type) {
+        const question1 = {
+            type: 'confirm',
+            name: 'provideEmail',
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/util/cli.js:65]: If user wants to set a note, but an email address is not in the notification, I ask if he wants to set it
2023-08-24_08h28_20
If not - exit
If yes - ask for email addresses

File: lib/index.js [hunk 13]
```diff
@@ -21,6 +21,10 @@ const cache = require('./util/cache');
  * main class
  */
 class Mcdev {
+    static buObject = {};
+    static properties;
+    static cred;
+    static bu;
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/index.js:27]: I would need to either pass properties in parameter to another function, or have 
const properties = await config.getProperties(); twice, so thought it would be easier to record it in static class variable. Same for buObject, cred and bu (otherwise need to pass that as parameters)

File: test/type.query.test.js [hunk 14]
```diff
@@ -498,7 +498,7 @@ describe('type: query', () => {
             // check number of API calls
             assert.equal(
                 testUtils.getAPIHistoryLength(),
-                35,
+                57,
```
Comment (JoernBerkefeld) [inline test/type.query.test.js:501]: similar to above: that seems a bit excessive? have you checked whats causing the massive increase?

File: test/type.automation.test.js [hunk 15]
```diff
@@ -27,8 +27,8 @@ describe('type: automation', () => {
             const result = cache.getCache();
             assert.equal(
                 result.automation ? Object.keys(result.automation).length : 0,
-                4,
-                'only four automations expected'
+                15,
```
Comment (JoernBerkefeld) [inline test/type.automation.test.js:30]: ouch

File: lib/index.js [hunk 16]
```diff
@@ -991,6 +1018,26 @@ class Mcdev {
                         break;
                     }
                 }
+                case 'clearNotifications':
+                case 'updateNotifications': {
+                    {
+                        if (
+                            Object.prototype.hasOwnProperty.call(MetadataTypeInfo[type], methodName)
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/index.js:1017]: regarding this one: type-adapted one in ImportFile.js
updateNotifications for importFile does a retrieve, update and then deploy (similar to fixKeys). How do I reuse existing retrieve, and deploy logic in ImportFile and avoid circular dependency error? I tried that at first, but eventually moved this into index.js

File: lib/index.js [hunk 17]
```diff
@@ -991,6 +1018,26 @@ class Mcdev {
...
+                        ) {
+                            resultArr.push(...(await MetadataTypeInfo[type][methodName](keyArr)));
+                            if (resultArr.length > 0) {
+                                Util.logger.info(`Retrieving ${type} to have most recent changes`);
+                                const retriever = new Retriever(this.properties, this.buObject);
+                                await retriever.retrieve([type], resultArr, null, false);
+                            }
+                        } else {
+                            resultArr.push(
+                                ...(await this.#updateNotificationsGeneric(type, keyArr))
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/index.js]: updateNotificationsGeneric does both(clear and update, technically remove notif em address is also an update on notifications) for importFile. I check which option was set when the function was called in MetadataType.js createOrUpdate()
image

File: lib/index.js [hunk 18]
```diff
@@ -1155,6 +1201,93 @@ class Mcdev {
...
+        return result;
+    }
+    /**
+     * Updates notification email address field
+     *
+     * @param {TYPE.SupportedMetadataTypes} type limit execution to given metadata type
+     * @param {string[]} [keyArr] customerkey of the metadata
+     * @returns {Promise.<Object.<string, string[]>>} key: business unit name, value: list of affected item keys
+     */
+    static async #updateNotificationsGeneric(type, keyArr) {
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/index.js]: That's what I tried to do, but couldn't reuse existing logic that retrieves, updates and deploys importFiles from ImportFile.js(circular dependency error). updateNotifications for importFile does a retrieve, update and then deploy (similar to fixKeys). How do I reuse existing retrieve, and deploy logic in ImportFile and avoid circular dependency error? I tried that at first, but eventually moved this into index.js

For example, to retrieve I would need to import Retriever class. in ImportFile. Retriever uses MetadataTypeInfo, and MetadataTypeInfo imports ImportFile.  So it's ImportFile->Retriever->MetadataTypeInfo->ImportFile

File: lib/metadataTypes/MetadataType.js [hunk 19]
```diff
@@ -2097,6 +2104,64 @@ class MetadataType {
         }
         return keysForDeploy;
     }
+    /**
+     * helper function to get a list of keys where notification email address should be updated
+     *
+     * @param {TYPE.MetadataTypeMap} metadataMap metadata mapped by their keyField
+     * @returns {string[]} list of keys
+     */
+    static async getKeysToSetNotifications(metadataMap) {
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/metadataTypes/MetadataType.js:2113]: >>i dont understand what this method is good for in this main class. It seems to be checking a single field only. We, however, are seeking to update 1, 2, 3 or even 4 fields at a time, which also happen to have different field-names depending on the type.

This method is used only for importFile, that is why there is only one field --completionEmail

File: lib/index.js [hunk 20]
```diff
@@ -826,6 +835,20 @@ class Mcdev {
...
+                checkMetadataSupport = false;
+                break;
+            }
+            case 'clearNotifications': {
+                lang_past = 'cleared notifications for';
+                lang_present = 'clearing notifications for';
+                requireKeyOrLike = true;
+                checkMetadataSupport = false;
+                break;
+            }
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/index.js]: +

File: lib/metadataTypes/MetadataType.js [hunk 21]
```diff
@@ -2097,6 +2104,64 @@ class MetadataType {
...
+            ` ☇ skipping ${this.definition.type}: updateNotifications is not supported yet for ${this.definition.type}`
+        );
+        return [];
+    }
+    /**
+     * Abstract clearNotifications method that needs to be implemented in child metadata type
+     *
+     * @returns {string[]} returns keys of items where notification email address/notes were removed
+     */
+    static async clearNotifications() {
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/metadataTypes/MetadataType.js]: done

File: test/type.automation.test.js [hunk 22]
```diff
@@ -883,4 +887,121 @@ describe('type: automation', () => {
...
+                'returned metadata was not equal expected for update'
+            );
+            assert.equal(
+                testUtils.getAPIHistoryLength(),
+                17,
+                'Unexpected number of requests made. Run testUtils.logAPIHistoryDebug() to see the requests'
+            );
+            return;
+        });
+        it('Should NOT update run error email address', async () => {
```
[the remark this answers is not in the document]
Comment (phjulia) [reply test/type.automation.test.js]: Yes, renamed

File: test/type.query.test.js (hunk 14 above)
Comment (phjulia) [reply test/type.query.test.js]: Yes, 14 new  automations , automation is a dependent type for query so it is retrieved, 2 requests per each so now should be 35+28=63

File: test/type.query.test.js [hunk 23]
```diff
@@ -458,7 +458,7 @@ describe('type: query', () => {
             // check number of API calls
             assert.equal(
                 testUtils.getAPIHistoryLength(),
-                33,
+                55,
```
[the remark this answers is not in the document]
Comment (phjulia) [reply test/type.query.test.js]: Yes, 14 new automations , automation is a dependent type for query so it is retrieved, 2 requests per each so now should be 35+28=63

File: test/type.automation.test.js [hunk 24]
```diff
@@ -883,4 +887,121 @@ describe('type: automation', () => {
...
+            );
+            assert.equal(
+                updatedNotificationsNote['testInstance/testBU'].length,
+                0,
+                'zero automation keys expected'
+            );
+            return;
+        });
+        it('Should clear all notification email addresses', async () => {
+            handler.setOptions({ clear: 'all' });
```
[the remark this answers is not in the document]
Comment (phjulia) [reply test/type.automation.test.js:1329]: 4 possible values: all, completionEmail, errorEmail and notes. Added 3 more tests (4 in total for clearNotif). Added info in cli in the description

File: lib/util/util.js [hunk 25]
```diff
@@ -198,6 +198,16 @@ const Util = {
...
         }
     },
+    /**
+     * helper that validates email address
+     *
+     * @param {string} email email to validate
+     * @returns {boolean} first elem is type, second elem is subType
+     */
+    emailValidator(email) {
+        const regex = new RegExp('[a-z0-9]+@[a-z]+.[a-z]{2,3}');
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/util/util.js]: I used [RFC 2822](https://www.rfc-editor.org/rfc/rfc2822#section-3.4.1) standard (the recommended one) that matches 99.99% of email addresses
https://stackoverflow.com/questions/46155/how-can-i-validate-an-email-address-in-javascript

File: lib/util/util.js [hunk 26]
```diff
@@ -198,6 +198,18 @@
...
     },
+    /**
+     * helper that validates email address
+     *
+     * @param {string} email email to validate
+     * @returns {boolean} first elem is type, second elem is subType
+     */
+    emailValidator(email) {
+        const regex = new RegExp(
+            "^[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$"
```
[the remark this answers is not in the document]
Comment (JoernBerkefeld) [reply lib/util/util.js:209]: can we check if there is a better one? (same for the other alert)

File: lib/metadataTypes/MetadataType.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (JoernBerkefeld) [reply lib/metadataTypes/MetadataType.js:2106]: ok then lets please relocate it to importFile class

File: lib/index.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (JoernBerkefeld) [reply lib/index.js]: as discussed in our call: 
let's reuse `index.js > #fixKeys()` method by making it more generic (similar to runOnBu / runMethod) because updateNotification and fixKeys are kinda doing similar things, checking & updating only specific files during deploy

File: lib/metadataTypes/MetadataType.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/metadataTypes/MetadataType.js:2106]: Moved to ImportFile

File: lib/index.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/index.js]: Created a function #updateItems() that does a retrieve and then deploy. Specific options to fixKeys and updateNotifications are set in #fixKeys and #updateNotifications

File: lib/util/cli.js [hunk 27]
```diff
@@ -53,7 +53,33 @@ const Cli = {
...
+        };
+        try {
+            if (!(await inquirer.prompt(question1)).provideEmail) {
+                return;
+            }
+            return (await inquirer.prompt(question2)).updateNotificationNotes;
+        } catch (ex) {
+            Util.logger.info(ex);
+        }
+    },
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/util/cli.js:82]: What kind of other scenarios do you mean? 

In case of an array, not sure how do I exit after asking first question and exiting if the user selects No? If they don't want to set an email address, I don't want to ask for email addresses (2 question)

>>alternatively, you could of course simply require inquirer in whatever class you need these questions and then add this logic there. would allow you to have more on-point questions which will increase UX
But then I would either have to have that in a separate function anyway because I ask these questions twice - if user wants to set errorNote and if they want to set completionNote, but no email addresses were provided

File: lib/util/cli.js [hunk 28]
```diff
@@ -53,7 +53,33 @@ const Cli = {
...
+     * interactive helper to set automation run completion/error note
+     *
+     * @param {TYPE.SupportedMetadataTypes} type note type (error/completion)
+     * @returns {Promise.<string[]>} responses
+     */
+    async updateNotificationEmails(type) {
+        const question1 = {
+            type: 'confirm',
+            name: 'provideEmail',
+            message: `To set run ${type} note, an email address should be set. Do you want to set it?`,
```
[the remark this answers is not in the document]
Comment (phjulia) [reply lib/util/cli.js:66]: ${type} is either completionNote or ErrorNote, I'll change it to ${noteType} to make it clearer

File: lib/util/cli.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (JoernBerkefeld) [reply lib/util/cli.js]: actually, this comment i should have deleted after our call :D sorry about that. you already clarified that you needed that exit and I honestly didn't read that into the code the first time around.

File: lib/util/cli.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (JoernBerkefeld) [reply lib/util/cli.js]: i meant the "To set run" part. That doesn't sound like proper English. one option could be to simply remove the word "run"
